In [1]:
import sentence_transformers

Link Transformers Model :  
https://huggingface.co/Snowflake/snowflake-arctic-embed-m

In [2]:
# MODEL INITIALIZATION

model = sentence_transformers.SentenceTransformer(model_name_or_path='Snowflake/snowflake-arctic-embed-m')

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


In [3]:
model , type(model) , len(model)

(SentenceTransformer(
   (0): Transformer({'max_seq_length': 512, 'do_lower_case': False}) with Transformer model: BertModel 
   (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': True, 'pooling_mode_mean_tokens': False, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
   (2): Normalize()
 ),
 sentence_transformers.SentenceTransformer.SentenceTransformer,
 3)

## compare words one by one

In [4]:
# ADD SENTENCES BAHASA INDONESIA
sentences1 = 'Dia merasa sangat senang ketika mendengar kabar baik itu'
sentences2 = 'Kabar baik tersebut membuatnya merasa sangat bahagia dan lega.'

In [5]:
# ADD SENTENCES ENGLISH

sentences3 = 'She felt very happy when she heard the good news.'
sentences4 = 'The good news made her feel incredibly joyful and relieved.'

In [6]:
embeddings1 = model.encode(sentences1, convert_to_tensor=True)
embeddings2 = model.encode(sentences2, convert_to_tensor=True)

embeddings3 = model.encode(sentences3, convert_to_tensor=True)
embeddings4 = model.encode(sentences4, convert_to_tensor=True)

embeddings1.shape , embeddings2.shape

(torch.Size([768]), torch.Size([768]))

In [7]:
# COMPARE 2 SENTENCES

similarity_indo = model.similarity(embeddings1, embeddings2)
similarity_eng = model.similarity(embeddings3, embeddings4)

print(f'Similarity between the sentences indonesia: {similarity_indo}')
print(f'Similarity between the sentences english: {similarity_eng}')

# COMPARE 2 SENTENCES USING COSINE SIMILARITY
from sklearn.metrics.pairwise import cosine_similarity

# RESHAPE TO 2D VECTOR
embeddings1 = embeddings1.reshape(1, 768)
embeddings2 = embeddings2.reshape(1, -1)

embeddings3 = embeddings3.reshape(1, -1)
embeddings4 = embeddings4.reshape(1, -1)

cosine_similar = cosine_similarity(embeddings1, embeddings2) # COSINE SIMILARITY EXPECTS 2D ARRAY
print(f'Cosine Similarity : {cosine_similar}')

cosine_similar2 = cosine_similarity(embeddings3, embeddings4)
print(f'Cosine Similarity : {cosine_similar2}')


Similarity between the sentences indonesia: tensor([[0.8449]])
Similarity between the sentences english: tensor([[0.9533]])
Cosine Similarity : [[0.8449221]]
Cosine Similarity : [[0.9532826]]


## compare sentences at once

In [8]:

sentences_indo = ['Dia merasa sangat senang ketika mendengar kabar baik itu.',
'Kabar baik tersebut membuatnya merasa sangat bahagia dan lega.',
'Setiap kali mendengar musik favoritnya, dia merasa penuh semangat.',
'Musik yang ia dengar itu membangkitkan perasaan bahagia yang mendalam.',
'Dia merasa nyaman berada di rumah setelah perjalanan panjang yang melelahkan.']

sentences_eng = ['She felt very happy when she heard the good news.',
'The good news made her feel incredibly joyful and relieved.',
'Every time she listens to her favorite music, she feels energized.',
'The music she hears stirs up a deep sense of happiness.',
'She felt comfortable being at home after the long and tiring trip.']

In [9]:
# ENCODE THEM

embeddings_indo = model.encode(sentences_indo)
embeddings_eng = model.encode(sentences_eng)

embeddings_indo , embeddings_indo.shape  , embeddings_eng , embeddings_eng.shape

(array([[-0.05329728,  0.05110858,  0.01709981, ..., -0.00337702,
          0.03150548,  0.04956917],
        [-0.03018753,  0.026941  ,  0.02517184, ..., -0.00911751,
          0.01815518,  0.04994358],
        [-0.05384601,  0.05427687, -0.007897  , ...,  0.01951372,
          0.0211255 ,  0.01155812],
        [-0.03230422,  0.07799845,  0.01277795, ...,  0.01963823,
          0.03129219,  0.01444418],
        [-0.06244443,  0.06240311,  0.00924083, ..., -0.01079868,
          0.02257819,  0.02732821]], dtype=float32),
 (5, 768),
 array([[-0.04619446,  0.01815579,  0.01900399, ...,  0.00713332,
          0.05034873,  0.01453644],
        [-0.0608112 ,  0.02716973,  0.02785231, ..., -0.00279166,
          0.062392  ,  0.0105258 ],
        [-0.02045596,  0.0112372 , -0.01580059, ...,  0.02191771,
          0.06453312, -0.01301301],
        [-0.06091132,  0.05548193,  0.00368568, ...,  0.02315031,
          0.05936863, -0.00109541],
        [-0.03718376,  0.00464372,  0.0340375 , ..., -

In [10]:
# COMPARE THEM

similarity_indo = model.similarity(embeddings_indo, embeddings_indo)
similarity_eng = model.similarity(embeddings_eng, embeddings_eng)

print(f'Similarity between the sentences:\n{similarity_indo}')
print(f'Similarity between the sentences:\n{similarity_eng}')

Similarity between the sentences:
tensor([[1.0000, 0.8838, 0.8645, 0.8019, 0.8658],
        [0.8838, 1.0000, 0.8526, 0.8378, 0.8875],
        [0.8645, 0.8526, 1.0000, 0.8893, 0.8592],
        [0.8019, 0.8378, 0.8893, 1.0000, 0.8203],
        [0.8658, 0.8875, 0.8592, 0.8203, 1.0000]])
Similarity between the sentences:
tensor([[1.0000, 0.9533, 0.8495, 0.8931, 0.8793],
        [0.9533, 1.0000, 0.8695, 0.8963, 0.8888],
        [0.8495, 0.8695, 1.0000, 0.9078, 0.8384],
        [0.8931, 0.8963, 0.9078, 1.0000, 0.8437],
        [0.8793, 0.8888, 0.8384, 0.8437, 1.0000]])


## Conclusion :

for indonesia Sentences :<br>
Sentence 1 and sentence 2 have a fairly high similarity value, which is 0.8838, indicating that the two sentences are very similar to each other. <br>
Sentence 1 and sentence 3 also have a fairly high similarity (0.8645), although slightly lower than sentence 1 and sentence 2. <br>
Sentence 3 and sentence 4 have a fairly high similarity value (0.8893), indicating that they are very similar. <br>
Sentence 4 and sentence 5 have a fairly high similarity (0.8203), although slightly lower than the other pairs.

for english sentences : <br>
Sentence 1 and sentence 2 have a very high similarity (0.9533), indicating that the two sentences are very similar to each other. <br>
Sentence 1 and sentence 4 have a fairly high similarity (0.8931), indicating that these two sentences are also very similar in the context of happiness and positive feelings.  <br>
Sentence 1 and sentence 3 are slightly lower with a similarity value of 0.8495, but still show a strong relationship related to the positive feelings created by music. <br>
Sentence 2 and sentence 4 have a relatively high similarity (0.8963), indicating that they are very similar in the context of happiness and comfort.